src/libfsm/dfavm.c

```diff
@@ -0,0 +1,1692 @@
+/*
+ * Copyright 2019 Shannon F. Stewman
+ *
+ * See LICENCE for the full copyright terms.
+ */
+
+#include <stdio.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <assert.h>
+#include <ctype.h>
+#include <errno.h>
+#include <string.h>
+
+#include <fsm/fsm.h>
+#include <fsm/vm.h>
+#include "internal.h"
+#include "print/ir.h"
+
+#define DEBUG_ENCODING     0
+#define DEBUG_VM_EXECUTION 0
+#define DEBUG_VM_OPCODES   0
```

Comment:
I'm not sure about the encoding and execution, but I do think `DEBUG_VM_OPCODES` should be exposed to both `re(1)` and `fsm(1)` as a `-l ...` output language option. (You can treat the `fsm_print` function pointer essentially as an enum for that)

Reply:
That makes sense.

I'm thinking we probably want to split things so there's a high-level "IR-ish" set of opcodes and a bytecode encoding.  The opcodes would be useful for optimization and would be the basis for any generated bytecode and assembly.

In this scheme, you'd want `fsm_print` functions for both the opcodes and bytecodes.